### PR TITLE
CQ4345796 date picker

### PR DESCRIPTION
**Pull request by s1mahanty**
Date Picker Authoring with Cypress Tests

## Description
Tasks Completed:
1. A minor Bug fix in TextInputImpl.java (earlier, getexclusiveMaximum() was returning exclusiveMinimum)
2. Modified Edit Dialog Date Picker
3. Added Authoring Cypress Test Cases 
4. Documentation (added README.md)
5. Core Component Example

Remaining feature: 

1. should switch to a custom date widget when patterns are added

## Related Issue
[CQ-4345796](https://jira.corp.adobe.com/browse/CQ-4345796)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml [hunk 1]
```diff
@@ -35,24 +37,48 @@
...
+                            <default
+                                    jcr:primaryType="nt:unstructured"
+                                    sling:orderBefore="guideGlobalLazyField"
+                                    sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                                    fieldDescription="Please enter the date in the required format  &quot;yyyy-mm-dd&quot;."
                                     fieldLabel="Default Value"
+                                    emptyText="YYYY-MM-DD"
                                     name="./default"/>
-                            <bindref
+                            <minimumMessage
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml:76]: These have to go in the validation tab.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml [hunk 2]
```diff
@@ -35,24 +37,48 @@
...
+                                    name="./placeholder"/>
+                            <default
+                                    jcr:primaryType="nt:unstructured"
+                                    sling:orderBefore="guideGlobalLazyField"
+                                    sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                                    fieldDescription="Please enter the date in the required format  &quot;yyyy-mm-dd&quot;."
                                     fieldLabel="Default Value"
+                                    emptyText="YYYY-MM-DD"
                                     name="./default"/>
-                            <bindref
```
Comment (vdua) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml:44]: also most of these properties would be coming from base.

**Review by vdua — CHANGES_REQUESTED** [review]

**Review by vdua — CHANGES_REQUESTED** [review]

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml [hunk 3]
```diff
@@ -15,146 +15,115 @@
...
                             sling:resourceType="granite/ui/components/coral/foundation/container"
                             active="{Boolean}true"
                             selected="{Boolean}true">
                         <items jcr:primaryType="nt:unstructured">
-                            <value
-                                    wrapperClass="cmp-adaptiveform-datepicker__value"
+                            <default
                                     jcr:primaryType="nt:unstructured"
-                                    sling:resourceType="granite/ui/components/coral/foundation/form/datepicker"
+                                    sling:orderBefore="guideGlobalLazyField"
```
[the remark this answers is not in the document]
Comment (s1mahanty) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml:42]: No we don't have it in base dialog.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ELEMENT cmp-adaptiveform-datepicker__errormessage
+```
+
+## JavaScript Data Attribute Bindings
+
+Apply a `data-cmp-is="adaptiveFormDatePicker"` attribute to the wrapper block to enable initialization of the JavaScript component.
+
+The following attributes can be added to the same element to provide options:
+
+1. `data-cmp-required-message` - populated with requiredMessage from the component's edit dialog
```
[the remark this answers is not in the document]
Comment (s1mahanty) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md]: Removed, Thanks!

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vdua) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml:42]: can you please remove `sling:orderBefore=guideGlobalLazyField`

**Review by nit23uec — APPROVED** [review]

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+2. `./hideTitle` - if set to `true`, the label of this field will be hidden
+3. `./name` - defines the name of the field, which will be submitted with the form data
+4. `./default` - defines the default value of the field
+5. `./description` - defines a help message that can be rendered in the field as a hint for the user
+6. `./required` - if set to `true`, this field will be marked as required, not allowing the form to be submitted until the field has a value
+7. `./requiredMessage` - defines the message displayed as tooltip when submitting the form if the value is left empty
+8. `./readOnly` - if set to `true`, the filed will be read only
+9. `./dataFormat` - defines the format in which the value exported/submitted
+10. `./editFormat` - defines the format in which the value will be edited by the user
+11. `./minimumDate` - define the minimum date input allowed for the field
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md:41]: this will need an update to capture all possible props that are actually being captured through edit dialog.
lets do it in another PR.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml [hunk 6]
```diff
@@ -15,146 +15,115 @@
...
                             jcr:title="Accessibility"
+                            granite:class="cmp-adaptiveform-datepicker__editdialog"
                             sling:resourceType="granite/ui/components/coral/foundation/container">
                         <items jcr:primaryType="nt:unstructured"/>
                     </accessibility>
                     <patterns
-                            jcr:title="Patterns"
                             jcr:primaryType="nt:unstructured"
+                            jcr:title="Formats"
+                            granite:class="cmp-adaptiveform-datepicker__editdialog"
```
Comment (nit23uec) [inline ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/_cq_dialog/.content.xml:109]: this granite:class should be removed from individual dialog fields.

File: ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md (hunk 5 above)
Comment (s1mahanty) [reply ui.apps/src/main/content/jcr_root/apps/core/fd/components/form/datepicker/v1/datepicker/README.md:41]: Sure.
